python/ray/util/retry.py

```diff
@@ -0,0 +1,38 @@
+"""Utils on retry."""
+
+import time
+from functools import wraps
+
+_RAY_DEFAULT_MAX_RETRY_COUNT = 10
+_RAY_DEFAULT_INIT_DELAY_SEC = 1
+_RAY_DEFAULT_MAX_DELAY_SEC = 30
+_RAY_DEFAULT_BACKOFF = 2
+_RAT_DEFAULT_JITTER_SEC = 1
+
+
+def retry(
+    max_retry_count=_RAY_DEFAULT_MAX_RETRY_COUNT,
+    init_delay_sec=_RAY_DEFAULT_INIT_DELAY_SEC,
+    max_delay_sec=_RAY_DEFAULT_MAX_DELAY_SEC,
+    backoff=_RAY_DEFAULT_BACKOFF,
+    jitter_sec=_RAT_DEFAULT_JITTER_SEC,
+):
```

Comment:
Do you know where should I place the unit test? I don't see `test` folder nearby

Reply:
this `retry` thing is only used in `release` at that one place, so can you just leave it in the file that uses it and do not put it in `ray` package?

the unit test for the `cluster_manager` file is here:
https://github.com/ray-project/ray/blob/master/release/BUILD.bazel#L422

Reply:
> this retry thing is only used in release at that one place

Yeah, that's true; but I somehow feel it could be used elsewhere in the future, considering it's so common to have something to retry.

I would create a new folder under `util`, so in the future all unit test on utils go there.
Let me know if you don't like it :)

Reply:
Unit tested added.

Reply:
so...


maybe we should talk on monday or something..

- firstly, high level, in general, one should assume "You aren't gonna need it" and limit the visibility as much as possible. https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it . if we need to reuse it in the future, it is easy to make a move a make a copy when we actually want to reuse it.
- secondly, files under `release/ray_release` directory is the ray release test wrapper. it is built as a stand-alone package, and does not depend on ray (there does not require `pip install ray` to work), so the `retry.py` here should not put here for that usage. `python/ray/util` will be packaged into `ray` wheels, that is an unnecessary additional exposure of unused code to ray users.
- if you really want to make it usable, put it under `release/ray_release/retry.py` or into `release/ray_release/util.py` ?

Reply:
> it is built as a stand-alone package, and does not depend on ray (it does not require pip install ray to work

https://buildkite.com/ray-project/premerge/builds/31058#019337be-867e-4fc4-b592-824e1f929966/186-553

````
[2024-11-17T01:43:23Z] ModuleNotFoundError: No module named 'ray'
````

Reply:
> if we need to reuse it in the future, it is easy to make a move a make a copy when we actually want to reuse it.

I don't quite agree with your argument.
The least visibility is to make the `retry` function as just a function inside of the `release.py` file;
but the problem is, without putting it into a centralized place, people are less likely to know about it and reuse.
A common consequence is everybody makes his/her own wheels for retry.

Reply:
IMO, if a function is independent of application logic, and the implementation/interface is general enough for reuse, it should be placed under a common folder for better visibility.

Reply:
> it is built as a stand-alone package, and does not depend on ray (it does not require pip install ray to work

I'm quite confused on the argument. Does that mean, if I want to use the same retry logic in runtime env agent and release package, I have to write two identical functions?

Reply:
> I have to write two identical functions?

it is recommended for this case.

[A little copying is better than a little dependency.](https://www.youtube.com/watch?v=PAAkCSZUG1c&t=9m28s)

https://www.youtube.com/watch?v=PAAkCSZUG1c&t=568s

or if there is a `retry` package that is secure, you can also import it I guess..